### PR TITLE
Implement more traits for guard types

### LINKED ISSUE #24372: Guard types should implement more traits
Types like `std::sync::MutexGuard` are semantically just a `&mut` with a destructor, and so it makes sense for them to implement the same sort of traits, e.g. `Display` and `Debug`.

### PR DESCRIPTION
**Pull request by marcusbuffett**
Implements `Hash`, `PartialEq`, `Eq`, `PartialOrd`, and `Ord` for
`Ref`, `RefMut`, `MutexGuard`, `RwLockReadGuard`, `RwLockWriteGuard`.

Fixes #24372

Bit of a rust noob, quick questions:

Preferred style wrt `x.deref()` or `*x`? I went with `*` unless it was going to be something too weird like `&**`.

`hash::Hash` or `Hash`? I see that `fmt::Display` and `fmt::Debug` are used for the Display and Debug traits, should I prefer that syntax instead?

I saw all the other implementations use `#[inline]` and I just followed suit, any rules of thumb here, or is it just "use inline everywhere always?"

I didn't see anything in the issue that said exactly what types needed these traits, so I just went with all the types that got `Debug` and `Display` in this commit: 14df54989aecb437e9cf6f3e066553cfbc979e26, any more types that need these traits?

### CODE REVIEW

File: src/libcore/cell.rs [hunk 1]
```diff
@@ -1131,6 +1187,61 @@ impl<'a, T: ?Sized + fmt::Display> fmt::Display for RefMut<'a, T> {
...
+
+#[stable(feature = "std_guard_impls_ext", since = "1.23.0")]
+impl<'a, T: ?Sized + Ord> Ord for RefMut<'a, T> {
+    #[inline]
+    fn cmp(&self, other: &RefMut<T>) -> Ordering {
+        self.value.cmp(&other.value)
+    }
+}
+
+#[stable(feature = "std_guard_impls_ext", since = "1.23.0")]
```
Comment (estebank) [inline src/libcore/cell.rs]: I believe these should be marked as `unstable` with a tracking issue set, but I'll let anyone else in the team correct me.

Comment (Mark-Simulacrum) [reply src/libcore/cell.rs]: We don't track stability on impls of traits today so these are all insta-stable AFAIK; no need to change anything.

Comment (kennytm) [inline src/libcore/cell.rs]: Unless you expect this PR will stay open for 20 days, the `since` field should still be 1.22.0.

**Comment by BurntSushi** [comment]
This all seems fine to me, but I'm kind of surprised we didn't already have these impls. @marcusbuffett Could you maybe elaborate on the use case for wanting each of these impls for these types?

cc @rust-lang/libs Is there a reason why we don't have these impls already or is it just that nobody has added them yet?

**Comment by marcusbuffett** [comment]
@BurntSushi Beyond pointing to the issue (#24372), I can't really elaborate much – I just searched for `E-easy` && `E-Help Wanted` tags and chose one. I don't know enough yet about the types to say why the impls should or shouldn't exist. I'd just be guessing, and your guess is better than mine.

File: src/libcore/cell.rs (hunk 1 above)
Comment (marcusbuffett) [reply src/libcore/cell.rs]: Done, thanks!

**Comment by alexcrichton** [comment]
Seems like a find idea to me! I think we just haven't gotten around to this previously

**Review by dtolnay — APPROVED** [review]
Seems okay to me. Thanks!

**Comment by alexcrichton** [comment]
Ok we got a chance to discuss this during the libs triage meeting today. In discussing this we ended up being increasingly wary of the guideline we set forth to add these trait impls and ended up concluding that we may not wish to add these at this time.

These implementations, as added here, wouldn't enable something like `MutexGuard<i32> == i32`, but that's arguably something we'd like to work. Adding that, however, precludes `MutexGuard<i32> == MutexGuard<i32>`. We didn't believe there was a clear way to implement this right now as all current strategies had downsides.

The possibility of specialization in the future may end up providing a solution to this problem, and @aturon also mentioned that the lang team is going to be working during the impl period on "deref coercion ergonomics" to alleviate much of the need for these impls in the first place.

In light of all that I'm going to close this for now and we'll likely be removing the guideline indicating that these traits should be implemented temporarily. Regardless though thanks for the PR @marcusbuffett!

**Pull request closed without merging**